src/scheduler/types/Scheduler.py

```diff
@@ -377,18 +382,120 @@ def bucket_stats(vals):
+                        else:
+                            options_for_stem_ = list(reversed(options_for_stem_))
+                            curr_best = None
+                            for ii, (prev_goodness, prev_budget, info) in enumerate(prev_curve):
+                                if ii != len(prev_curve) - 1 and len(options_for_stem_) > 0:
+                                    next_pt = prev_curve[ii+1]
+                                    next_pt = (agg_func(next_pt[0], options_for_stem_[0][0]), next_pt[1] + options_for_stem_[0][1])
+                                else:
+                                    next_pt = None
+                                for c_benefit, c_cost, c_unit in options_for_stem_:
+                                    ops += 1
+                                    new_budget = prev_budget + c_cost
+                                    # Pruning
+                                    if new_budget > cost_threshold:
+                                        # Note: break depends on reversed, otherwise must be continue.
+                                        break
+                                    new_goodness = agg_func(prev_goodness, c_benefit)
+                                    if curr_best is not None and curr_best >= (new_goodness, -new_budget):
+                                        continue
+                                    if next_pt is not None and next_pt[1] <= new_budget and next_pt[0] >= new_goodness:
+                                        break
+                                    curr_best = (new_goodness, -new_budget)
+                                    min_objective_by_budget.append((new_goodness, new_budget, {'unit': c_unit, 'prev': info}))
+                                    updates += 1
+
+                            min_objective_by_budget = list(reversed(min_objective_by_budget))
+                        dp[i] = scheduler_util.make_monotonic(min_objective_by_budget)
+
+                    stem_sols = [(goodness, budget, info)
+                                 for goodness, budget, info in dp[len(self.apps) - 1]
+                                 if budget <= cost_threshold]
+                    if len(stem_sols) > 0:
+                        best_stem_sol = stem_sols[0]
+                        if best_result is None or best_result[0] < best_stem_sol[0] or (best_result[0] == best_stem_sol[0] and best_result[1] > best_stem_sol[1]):
+                            if self.verbose > 0:
+                                print 'improved:', stem, best_stem_sol
+                            stems_improved += 1
+                            best_result = best_stem_sol
+            # constraints, total stems, total stems in budget, total stems that produced more optimal solutions, solutions, ops
+            print k, num_stems, num_stems_in_budget, stems_improved, len(solutions), ops, ops / max(1, num_stems_in_budget), updates
```

Comment:
Hard to parse in stdout. Could be better in the verbose or to add more context in the print statement